src/huggingface_hub/utils/_verification.py

```diff
@@ -0,0 +1,176 @@
+import re
+from dataclasses import dataclass
+from pathlib import Path
+from typing import TYPE_CHECKING, Literal, Optional, TypedDict, Union
+
+from .. import constants
+from ..file_download import repo_folder_name
+from .sha import git_hash, sha_fileobj
+
+
+if TYPE_CHECKING:
+    from ..hf_api import RepoFile, RepoFolder
+
+# using fullmatch for clarity and strictness
+_REGEX_COMMIT_HASH = re.compile(r"^[0-9a-f]{40}$")
+
+
+# Typed structure describing a checksum mismatch
+class Mismatch(TypedDict):
+    path: str
+    expected: str
+    actual: str
+    algorithm: str
+
+
+HashAlgo = Literal["sha256", "git-sha1"]
+
+
+@dataclass(frozen=True)
+class FolderVerification:
+    revision: str
+    checked_count: int
+    mismatches: list[Mismatch]
+    missing_paths: list[str]
+    extra_paths: list[str]
+    verified_path: Optional[Path] = None
```

Comment:
^here is what I meant above


```suggestion
    verified_path: Path
```